docs/src/main/tut/docs/stack/cache.md

```diff
@@ -0,0 +1,241 @@
+We should be able to check how much crates of some specific apple cultivar the company has in stock and finally to register an order.
+
+We will represent these capabilities using two algebras: `CustomerPersistence` and `StockPersistence`.
+
+```tut:book
+object algebras {
+  @free trait CustomerPersistence[F[_]] {
+    def getCustomer(id: CustomerId): FreeS.Par[F, Option[Customer]]
+  }
+
+  @free trait StockPersistence[F[_]] {
+    def checkQuantityAvailable(cultivar: String): FreeS.Par[F, Int]
+    def registerOrder(order: Order): FreeS.Par[F, Unit]
+  }
+}
+```
+
+These two algebras will be combined together in a `Persistence` module, that in its turn will be a part of our `App` module, that additionally includes the
+capabilities to fail, to read from our `Config` configuration and to cache `Customer`s using respectively `error` and `reader` from the _freestyle-effects_ module and `cache` from the _freestyle-cache_ module.
+
+```tut:book
+import freestyle.effects.error._
+import freestyle.effects.reader
+import freestyle.cache.KeyValueProvider
+
+object modules {
+  val rd = reader[Config]
+  val cacheP = new KeyValueProvider[CustomerId, Customer]
+
+  @module trait Persistence[F[_]] {
+    val customer: algebras.CustomerPersistence[F]
+    val stock: algebras.StockPersistence[F]
+  }
+
+  @module trait App[F[_]] {
+    val persistence: Persistence[F]
+
+    val errorM: ErrorM[F]
+    val cacheM: cacheP.CacheM[F]
+    val readerM: rd.ReaderM[F]
+  }
+}
+```
+
+To validate the order we check if the number of crates that is ordered bigger is than zero and that the requested apple cultivar is selled by the company.
+
+For the second part we use the `reader` effect algebra to read the set of apple cultivars from our `Config`.
+
+```tut:book
+import modules._
+import cats.data.{NonEmptyList, ValidatedNel}
+import cats.implicits._
+
+def validateOrder[F[_]](order: Order, customer: Customer)(implicit app: App[F]): FreeS.Par[F, ValidatedNel[String, Unit]] =
+  app.readerM.reader { config =>
+    val v = ().validNel[String]
+    v.ensure(NonEmptyList.of(
+      "Number of crates ordered should be bigger than zero"))(
+      _ => order.crates > 0) |+|
+    v.ensure(NonEmptyList.of(
+      "Apple cultivar is not available"))(
+      _ => config.cultivars.contains(order.cultivar.toLowerCase))
+  }
+```
+
+When validating an order we need to get the information of the customer, if a customer places multiple orders we don't want to send a database request every time, so we could use the `freestyle-cache` module.
+
+In the `getCustomer` function we try to find the customer in the cache, otherwise falling back to getting the customer from a database (or other persistence store).
+
+```tut:book
+import cats.data.OptionT
+
+def getCustomer[F[_]](id: CustomerId)(implicit app: App[F]): FreeS[F, Option[Customer]] =
+  // first try to get the customer from the cache
+  OptionT(app.cacheM.get(id).freeS).orElseF {
```

Comment:
Can we replace here `OptionT` with the `OptionM` effect algebra to avoid using a transformer? One of the nicest thigns about these effects algebras is that for most common cases they remove the need to use transformers in the program and logic declarations since the constrains are in the interpreters.